### PR TITLE
Better checkpointing/fault tolerance + tests

### PR DESCRIPTION
**Pull request by krfricke**
Better checkpoint handling. Instead of writing local checkpoints, this PR sends checkpoints from one worker via the queue to the driver. This is effectively checkpointing via the object store, thus no longer requiring local temp files.

It is not straightforward to load a checkpoint into Rabit, so this way seems to be the most versatile way to make sure each worker has access to the latest checkpoint.

Todos (possibly in follow-up PRs): 
- Multinode release test
- Save last checkpoint on driver on fail if maximum number of restarts is exceeded

### CODE REVIEW

File: xgboost_ray/main.py
Comment (richardliaw) [inline xgboost_ray/main.py:253]: should we use something more efficient or put it into the object store?

Comment (richardliaw) [reply xgboost_ray/main.py:253]: Does it also work on GPUs?

File: xgboost_ray/tests/utils.py
Comment (richardliaw) [inline xgboost_ray/tests/utils.py:65]: nit: keys = keys or []

File: xgboost_ray/main.py
Comment (richardliaw) [inline xgboost_ray/main.py:282]: is this an issue/bug?

File: xgboost_ray/main.py
Comment (richardliaw) [inline xgboost_ray/main.py:250]: Do you know why the xgboost internal callbacks use a Class-based callback?

https://github.com/dmlc/xgboost/blob/927c316aeb0a948e9635c1ffe54d5e6ed25f6797/python-package/xgboost/callback.py#L668-L677

File: xgboost_ray/main.py
Comment (richardliaw) [inline xgboost_ray/main.py:476]: maybe instead, do `_checkpoint.__dict__.update(item.__dict__)`

File: xgboost_ray/main.py
Comment (richardliaw) [inline xgboost_ray/main.py:386]: hmm, what does this do?

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:250]: This seems to be a new change, here is the legacy callback handler: https://github.com/dmlc/xgboost/blob/927c316aeb0a948e9635c1ffe54d5e6ed25f6797/python-package/xgboost/callback.py#L682

It will be added in the next release (1.3.0, current stable release 1.2.x)

It should be easy to refactor our callbacks in the future, but I'd like to keep them in the legacy format to support older installations of xgboost. We should put this on our backlog though and possibly update in (late) Q1/2021

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:386]: Basically we're passing a checkpoint and results reference to be able to record (and return) the contents even in the case of failure. Since we re-raise the actor error in the try/catch block in `_train()`, this seemed like the best way to support keeping checkpoint and results states, and since `_train()` is an internal function it should be fine to use this API here. 
I'm open to suggestions here, though.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:253]: It should work on GPUs, I'll try this though.

For the object store, should we `ray.put` the model into the object store and pass the reference via the queue?
Also, it would be great if you could explain why the current approach is inefficient, so that I gather some intuition about this. Thanks!

Comment (richardliaw) [reply xgboost_ray/main.py:253]: Yes! so from what i understand, the `model` is essentially a wrapper around some numerical datatype (array). Pickle serialization of numerical datatypes is often inefficient. Rather, the Plasma object store has much better suited for transferring numerical datatypes, so that's why `ray.put` is better! 

See: https://arrow.apache.org/blog/2017/10/15/fast-python-serialization-with-ray-and-arrow/

File: xgboost_ray/main.py
Comment (richardliaw) [reply xgboost_ray/main.py:250]: Cool sounds good!

**Review by richardliaw — APPROVED** [review]
Looks great!

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:282]: I think so. I'll check this tomorrow and write a small reproducible script, possibly open an issue on ray core

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:253]: Got it, thanks! I'll address this in a follow-up PR though to unblock the tune checkpointing

Comment (krfricke) [reply xgboost_ray/main.py:253]: Actually, I'll look into this before merging, it might be related to the error we currently see on the latest release

File: xgboost_ray/main.py
Comment (amogkam) [reply xgboost_ray/main.py:386]: Got it. So just for my understanding, `additional_results` and `num_boost_rounds` are not correctly being handled during failure in the current implementation correct? Without this PR, when a training run has to restart, each actor loads its local checkpoint, but it starts again from iteration 0 and previous `additional_results` are not being saved, is that right?

**Review by amogkam — APPROVED** [review]
LGTM! Just left one question to help with my understanding.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:386]: Yes, that is exactly right. Without this PR the restarted actors train for `num_boost_rounds` again and forfeit any additional results they collected before.

**Comment by krfricke** [comment]
Thanks, I'll look into the fail in the ray release sometime today/tomorrow before merging.

**Pull request merged**
